### PR TITLE
bugfix: future for empty http 2 response does not complete. fixes #4233

### LINKED ISSUE #4233: Future returned by Http2ServerResponse#end() does not complete.
### Version

4.2.3

### Context

Writing a simple HttpServer where i have to wait for completed Http 1 / 2 responses.

### Do you have a reproducer?

```
package com.github.tomakehurst.wiremock.it;

import io.vertx.core.CompositeFuture;
import io.vertx.core.Vertx;
import io.vertx.core.http.HttpClientOptions;
import io.vertx.core.http.HttpClientRequest;
import io.vertx.core.http.HttpMethod;
import io.vertx.core.http.HttpVersion;
import io.vertx.ext.unit.Async;
import io.vertx.ext.unit.TestContext;
import io.vertx.ext.unit.junit.VertxUnitRunner;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.Collections;

@RunWith(VertxUnitRunner.class)
public class ReproducerTest {

  @Test
  public void reproducer(final TestContext context) {
    final Vertx vertx = Vertx.vertx();
    final Async async = context.strictAsync(2);

    vertx
      .createHttpServer()
      .requestHandler(
        request -> request.response().setStatusCode(200).end().onComplete(result -> async.countDown())
      )
      .listen(8080)
      .compose(
        server -> CompositeFuture.all(
          vertx.createHttpClient(new HttpClientOptions()
              .setUseAlpn(true)
              .setProtocolVersion(HttpVersion.HTTP_2)
              .setAlpnVersions(Collections.singletonList(HttpVersion.HTTP_2))
              .setHttp2ClearTextUpgrade(false)
              .setDefaultHost("localhost")
              .setDefaultPort(8080)
            )
            .request(HttpMethod.GET, "uri")
            .compose(HttpClientRequest::send)
            .onSuccess(response -> context.assertEquals(200, response.statusCode()))
            .mapEmpty()
          ,
          vertx.createHttpClient(new HttpClientOptions()
              .setDefaultHost("localhost")
              .setDefaultPort(8080)
            )
            .request(HttpMethod.GET, "uri")
            .compose(HttpClientRequest::send)
            .onSuccess(response -> context.assertEquals(200, response.statusCode()))
            .mapEmpty()
        )
      )
      .onComplete(context.asyncAssertSuccess());
  }
}
```

### Extra

openjdk 17.0.1

### PR DESCRIPTION
**Pull request by DemonicTutor**
Signed-off-by: Markus Spika <markus.spika@gmail.com>

Motivation:

https://github.com/eclipse-vertx/vert.x/issues/4233

Conformance:

You should have signed the Eclipse Contributor Agreement as explained in https://github.com/eclipse/vert.x/blob/master/CONTRIBUTING.md
Please also make sure you adhere to the code style guidelines: https://github.com/vert-x3/wiki/wiki/Vert.x-code-style-guidelines

### CODE REVIEW

File: src/main/java/io/vertx/core/http/impl/Http2ServerResponse.java [hunk 1]
```diff
@@ -453,6 +453,8 @@ void write(ByteBuf chunk, boolean end, Handler<AsyncResult<Void>> handler) {
       boolean sent = checkSendHeaders(end && !hasBody && trailers == null);
       if (hasBody || (!sent && end)) {
         stream.writeData(chunk, end && trailers == null, handler);
+      } else {
+        handler.handle(Future.succeededFuture());
```
Comment (vietj) [inline src/main/java/io/vertx/core/http/impl/Http2ServerResponse.java:457]: the handler should be called outside of the synchronized block like the endHandler and the bodyEndHandler

**Comment by vietj** [comment]
can you add a test ?

**Comment by DemonicTutor** [comment]
@vietj ah of course! .. i assume the right place for the test would be io.vertx.core.http.HttpTest ?

**Comment by vietj** [comment]
yes, it sounds like it should be tested for HTTP/1 and HTTP/2

File: src/test/java/io/vertx/core/http/HttpTest.java [hunk 2]
```diff
@@ -6632,4 +6628,41 @@ public void testConnectTimeout() {
       }));
     await();
   }
+
+  @Test
+  public void testResponseEndFutureCompletes_WithoutBody() throws Exception {
+      testResponseEndFutureCompletes(HttpServerResponse::end);
```
Comment (DemonicTutor) [inline src/test/java/io/vertx/core/http/HttpTest.java:6634]: @vietj this is the one failing without the code change but i added the other cases as well to cover it explicitly

Comment (vietj) [reply src/test/java/io/vertx/core/http/HttpTest.java:6634]: ok

**Pull request merged**